homeassistant/components/powerwall/__init__.py

```diff
@@ -222,30 +222,28 @@ async def _login_and_fetch_base_info(
 async def _call_base_info(power_wall: Powerwall, host: str) -> PowerwallBaseInfo:
     """Return PowerwallBaseInfo for the device."""
 
-    (
-        gateway_din,
-        site_info,
-        status,
-        device_type,
-        serial_numbers,
-    ) = await asyncio.gather(
-        power_wall.get_gateway_din(),
-        power_wall.get_site_info(),
-        power_wall.get_status(),
-        power_wall.get_device_type(),
-        power_wall.get_serial_numbers(),
-    )
-
-    # Serial numbers MUST be sorted to ensure the unique_id is always the same
-    # for backwards compatibility.
-    return PowerwallBaseInfo(
-        gateway_din=gateway_din.upper(),
-        site_info=site_info,
-        status=status,
-        device_type=device_type,
-        serial_numbers=sorted(serial_numbers),
-        url=f"https://{host}",
-    )
+    try:
+        async with asyncio.TaskGroup() as tg:
+            gateway_din = tg.create_task(power_wall.get_gateway_din())
+            site_info = tg.create_task(power_wall.get_site_info())
+            status = tg.create_task(power_wall.get_status())
+            device_type = tg.create_task(power_wall.get_device_type())
+            serial_numbers = tg.create_task(power_wall.get_serial_numbers())
+
+        # Serial numbers MUST be sorted to ensure the unique_id is always the same
+        # for backwards compatibility.
+        return PowerwallBaseInfo(
+            gateway_din=gateway_din.result().upper(),
+            site_info=site_info.result(),
+            status=status.result(),
+            device_type=device_type.result(),
+            serial_numbers=sorted(serial_numbers.result()),
+            url=f"https://{host}",
+        )
+    except ExceptionGroup as e:
+        raise e.exceptions[0] from None
+    except BaseExceptionGroup as e:
+        raise e.exceptions[0] from None
 
```

Comment:
Why catch the exception? Shouldn't we let it raise to work the same way as before?

Reply:
When it comes it comes to exceptions, the big difference between gather and TaskGroup is that gather just returns the first raised exception and TaskGroup returns them all wrapped in these group classes.   To mimic what gather does, I had to re-raise the first exception from the group.

The code that calls this assumes that it is only handling the first raised exception and isn't setup to handle the group, so I'm just replicating how the code worked before.

Reply:
It is just some really inconvenient side effect that type checking only extends up to 6 items for gather [here](https://github.com/python/typeshed/blob/main/stdlib/asyncio/tasks.pyi#L134).    🤷

Reply:
Would you please add a comment explaining whats going on here as it wasn't obvious at first glace.

I think `ExceptionGroup` likely inherits from `BaseExceptionGroup` so I'm not sure you need both.  Not at my desk to verify that though.

Reply:
Sure I can add a comment.

I guess strict type checking is a good thing but in cases like this it is a real pain in the ass.

Both those exceptions have different base classes and have really unintuitive names.  [The docs](https://docs.python.org/3/library/exceptions.html#BaseExceptionGroup) talk about the difference.

Their hierarchy is:
```
BaseException
 ├── BaseExceptionGroup
 └── Exception
      ├── ExceptionGroup [BaseExceptionGroup]
```

Reply:
Thinking about this some more, I think this mades everything less readable and harder to follow we should keep `asyncio.gather()` and add mypy ignore statements with a comment linking to the problem https://github.com/python/typeshed/blob/main/stdlib/asyncio/tasks.pyi?rgh-link-date=2024-01-28T17%3A32%3A27Z#L134   Maybe mypy will fix that in the future

Reply:
Oh I guess it is multiple inheritance so I can just handle BaseExceptionGroup

Reply:
> Thinking about this some more, I think this mades everything less readable and harder to follow and we should keep `asyncio.gather()`, and instead add mypy ignore statements with a comment linking to the problem https://github.com/python/typeshed/blob/main/stdlib/asyncio/tasks.pyi?rgh-link-date=2024-01-28T17%3A32%3A27Z#L134
> 
> Maybe mypy will fix that in the future

It is true that this makes it weird to read.   On the plus side, it is does maintain complete real type checking which is the point of having such a strict type checking regime.   I figured that in the long run, the value of type checking outweighs the weirdness here.

Reply:
Also TaskGroup does a way better job of gracefully handling corner cases as compared to gather.   It is a much better API for this case, just a little funky in the syntax.

Reply:
Its definitely not an easy call to make here. I'll sleep on this one.